cbtf: Update Boost dependency

hainest opened the pull request.

hainest committed:
cbtf: Update Boost dependency

hainest committed:
style update

hainest commented:
Derp. This will have to wait until #30627 is done.

tldahlgren reviewed (dismissed):
LGTM. What do you think @jgalarowicz ?

bernhardkaindl commented on var/spack/repos/builtin/packages/cbtf/package.py:
```diff
@@ -48,12 +47,9 @@ class Cbtf(CMakePackage):
-    # See https://github.com/spack/spack/pull/22303 for reference
-    depends_on(Boost.with_default_variants)
+    depends_on(
+        "boost@1.70.0:+date_time+filesystem+preprocessor+signals2+spirit_classic+test+thread"
```
The open-ended version range for boost does not work, it faild with boost-1.86.0.

If I remember correctly, path.is_complete() is version-specific:
```py
5 errors found in build log:
     111    In file included from opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/boost-1.86.0-vzz4cvt6zj7dcdodo4usj2ijz
            33hf2dl/include/boost/bind.hpp:30,
     112                     from .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/libcbtf/Re
            solvePath.cpp:21:
     113    opt/spack/linux-ubuntu22.04-skylake/gcc-12.3.0/boost-1.86.0-vzz4cvt6zj7dcdodo4usj2ijz33hf2dl/include/boost/
            bind.hpp:36:1: note: '#pragma message: The practice of declaring the Bind placeholders (_1, _2, ...) in the global namespace is d
            eprecated. Please use <boost/bind/bind.hpp> + using namespace boost:placeholders, or define BOOST_BIND_GLOBAL_PLACEHOLDERS to re
            tain the current behavior.'
     114       36 | BOOST_PRAGMA_MESSAGE(
     115          | ^~~~~~~~~~~~~~~~~~~~
     116    .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/libcbtf/ResolvePath.cpp: In func
            tion 'boost:filesystem:path KrellInstitute:CBTF:Impl:resolvePath(const std:vector<boost:filesystem:path>&, const boost:f
            ilesystem:path&)':
  >> 117    .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/libcbtf/ResolvePath.cpp:130:14:
            error: 'const class boost:filesystem:path' has no member named 'is_complete'
     118      130 |     if (path.is_complete())
     119          |              ^~~~~~~~~~~
     120    .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/libcbtf/ResolvePath.cpp: In func
            tion 'boost:filesystem:path KrellInstitute:CBTF:Impl:resolvePath(const FileType&, const boost:filesystem:path&)':
  >> 121    .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/libcbtf/ResolvePath.cpp:158:14:
            error: 'const class boost:filesystem:path' has no member named 'is_complete'
     122      158 |     if (path.is_complete())
     123          |              ^~~~~~~~~~~
  >> 124    make[2]: *** [libcbtf/CMakeFiles/cbtf.dir/build.make:107: libcbtf/CMakeFiles/cbtf.dir/ResolvePath.cpp.o] Error 1
     125    make[2]: Leaving directory '.stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/buil
            d_cbtf'
  >> 126    make[1]: *** [CMakeFiles/Makefile2:266: libcbtf/CMakeFiles/cbtf.dir/all] Error 2
     127    make[1]: Leaving directory '.stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-src/buil
            d_cbtf'
  >> 128    make: *** [Makefile:139: all] Error 2

See build log for details:
  .stage/spack-stage-cbtf-1.9.4.1-kdbudmyeo2wcwcqfmlpgl3ggsngj46eu/spack-build-out.txt
```

hainest replied:
>If I remember correctly, path.is_complete() is version-specific:

I'll check on that.

hainest replied:
`is_complete` was removed in 1.85.0, so I added a version constraint.

bernhardkaindl commented on var/spack/repos/builtin/packages/cbtf/package.py:
```diff
@@ -48,12 +47,9 @@ class Cbtf(CMakePackage):
-    # See https://github.com/spack/spack/pull/22303 for reference
-    depends_on(Boost.with_default_variants)
+    depends_on(
+        "boost@1.70.0:+date_time+filesystem+preprocessor+signals2+spirit_classic+test+thread"
```
There is no variant +preprocessor in boost and no other package requests such variant:
```py
grep -r +preprocessor var/spack/repos/builtin/packages/
var/spack/repos/builtin/packages/cbtf/package.py:        "boost@1.70.0:+date_time+filesystem+preprocessor+signals2+spirit_classic+test+thread"
```
The precheck in CI tells the same.

hainest replied:
Yup. Hence my [comment](https://github.com/spack/spack/pull/47131#issuecomment-2429299063).

hainest replied:
I've removed the "next-gen" boost variants so this can get merged. I need this PR to make testing easier for those changes.

hainest committed:
Remove next-gen boost variants

These will be added back after #30627 is merged.

hainest committed:
Limit Boost version

cbtf uses boost:filesystem:path::is_complete() that was removed in Boost v1.85.0.

bendamasio reviewed (commented):
Hello @jgalarowicz, I generated some build results:<br>
:information_source: `spack install` on the changed recipes of this PR was successful[^1] on Ubuntu 22.04.4 LTS!<details><summary>Click here for a summary of the successful builds.</summary>
These changed specs were found using [gh pr diff](https://cli.github.com):
- `cbtf@=1.9.4`
- `cbtf@=1.9.4.1`
- `cbtf@=1.9.3`


<details><summary><b>Click here to show/hide the report on </b><tt>spack find</tt> for <tt>linux-ubuntu22.04-skylake / gcc@13.1.0</tt>.</summary><ol>

<li><details><summary>Click on this area to unfold/refold the dependency report of this build:

>```py
>cbtf@1.9.3~ipo~runtime build_system=cmake build_type=RelWithDebInfo generator=make
>```
</summary>

>```py
>	boost@1.84.0atomicchrono_time_parallel_options_erasure cxxstd=11 visibility=hidden
>		bzip2@1.0.8
>			diffutils@3.10
>		xz@5.4.6
>		zlib-ng@2.2.1compatnew_strategiesopt
>		zstd@1.5.6programs
>	cmake@3.30.5
>		curl@8.10.1 tls=openssl
>			nghttp2@1.63.0
>			openssl@3.3.1 certs=mozilla
>				ca-certificates-mozilla@2023-05-30
>		ncurses@6.5
>	gcc-runtime@13.1.0
>	glibc@2.35
>	gmake@4.4.1
>	libtirpc@1.3.3
>		krb5@1.21.3
>			bison@3.8.2
>				m4@1.4.19sigsegv,bfdffa7
>					libsigsegv@2.14
>			findutils@4.9.0
>			perl@5.40.0cpanmopcodeopenthreads
>				berkeley-db@18.1.40cxx,b231fcc
>				gdbm@1.23
>					readline@8.2
>	libxml2@2.10.3
>		pkgconf@2.2.0
>	mrnet@5.0.1-3lwthreads
>	rpcsvc-proto@1.4.4
>		gettext@0.22.5bzip2cursesgit
>			tar@1.34 zip=pigz
>				pigz@2.8
>	xerces-c@3.2.5 netaccessor=curl transcoder=gnuiconv
>
>```

</details>


<li><details><summary>Click on this area to unfold/refold the dependency report of this build:

>```py
>cbtf@1.9.4~ipo~runtime build_system=cmake build_type=RelWithDebInfo generator=make
>```
</summary>

>```py
>	boost@1.84.0atomicchrono_time_parallel_options_erasure cxxstd=11 visibility=hidden
>		bzip2@1.0.8
>			diffutils@3.10
>		xz@5.4.6
>		zlib-ng@2.2.1compatnew_strategiesopt
>		zstd@1.5.6programs
>	cmake@3.30.5
>		curl@8.10.1 tls=openssl
>			nghttp2@1.63.0
>			openssl@3.3.1 certs=mozilla
>				ca-certificates-mozilla@2023-05-30
>		ncurses@6.5
>	gcc-runtime@13.1.0
>	glibc@2.35
>	gmake@4.4.1
>	libtirpc@1.3.3
>		krb5@1.21.3
>			bison@3.8.2
>				m4@1.4.19sigsegv,bfdffa7
>					libsigsegv@2.14
>			findutils@4.9.0
>			perl@5.40.0cpanmopcodeopenthreads
>				berkeley-db@18.1.40cxx,b231fcc
>				gdbm@1.23
>					readline@8.2
>	libxml2@2.10.3
>		pkgconf@2.2.0
>	mrnet@5.0.1-3lwthreads
>	rpcsvc-proto@1.4.4
>		gettext@0.22.5bzip2cursesgit
>			tar@1.34 zip=pigz
>				pigz@2.8
>	xerces-c@3.2.5 netaccessor=curl transcoder=gnuiconv
>
>```

</details>


<li><details><summary>Click on this area to unfold/refold the dependency report of this build:

>```py
>cbtf@1.9.4.1~ipo~runtime build_system=cmake build_type=RelWithDebInfo generator=make
>```
</summary>

>```py
>	boost@1.84.0atomicchrono_time_parallel_options_erasure cxxstd=11 visibility=hidden
>		bzip2@1.0.8
>			diffutils@3.10
>		xz@5.4.6
>		zlib-ng@2.2.1compatnew_strategiesopt
>		zstd@1.5.6programs
>	cmake@3.30.5
>		curl@8.10.1 tls=openssl
>			nghttp2@1.63.0
>			openssl@3.3.1 certs=mozilla
>				ca-certificates-mozilla@2023-05-30
>		ncurses@6.5
>	gcc-runtime@13.1.0
>	glibc@2.35
>	gmake@4.4.1
>	libtirpc@1.3.3
>		krb5@1.21.3
>			bison@3.8.2
>				m4@1.4.19sigsegv,bfdffa7
>					libsigsegv@2.14
>			findutils@4.9.0
>			perl@5.40.0cpanmopcodeopenthreads
>				berkeley-db@18.1.40cxx,b231fcc
>				gdbm@1.23
>					readline@8.2
>	libxml2@2.10.3
>		pkgconf@2.2.0
>	mrnet@5.0.1-3lwthreads
>	rpcsvc-proto@1.4.4
>		gettext@0.22.5bzip2cursesgit
>			tar@1.34 zip=pigz
>				pigz@2.8
>	xerces-c@3.2.5 netaccessor=curl transcoder=gnuiconv
>
>```

</details>



</details>

[^1]: Orchestrated using `spack install` by [github.com/spack/gh-spack-pr](https://github.com/spack/gh-spack-pr) and the [GitHub CLI](https://cli.github.com/)
</details><details><summary>Click here for a disclaimer for the maintainers of the recipes

</summary><br>👍 Hello @jgalarowicz!<br><br>

- We would like to ask you if you could review this PR within 5 days.
- If the PR is not reviewed within 5 days, the PR can be approved and merged by members.
- Please use 👍, Approve, LGTM or Request in the comment so we can act accordingly.
- If you like to review but need more time, please add a comment to the PR.
- If you have questions (or would like to be removed as maintainer), please let us know.

Thank you for your time and effort!<br>
You can also join the Spack Slack channel for more information and have a chat:
https://spackpm.slack.com → #pull-requests. Thank you and best regards, @x64-Linux-Builder</details>

bernhardkaindl reviewed (approved):
👍 @jgalarowicz: Your review was requested last week by @spackbot.

We are free to merge after 5-7 days of no further review, so, I'll enable auto-merge after CI pass.
